presto-hive/src/main/java/com/facebook/presto/hive/util/HiveFileIterator.java

```diff
@@ -0,0 +1,166 @@
+import org.apache.hadoop.mapred.InputFormat;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.List;
+import java.util.Properties;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class HiveFileIterator
+        extends AbstractIterator<LocatedFileStatus>
+{
+    private final FileSystem fileSystem;
+    private final DirectoryLister directoryLister;
+    private final NamenodeStats namenodeStats;
+    private final Path path;
+    private final String partitionName;
+    private final InputFormat<?, ?> inputFormat;
+    private final Properties schema;
+    private final List<HivePartitionKey> partitionKeys;
+    private final TupleDomain<HiveColumnHandle> effectivePredicate;
+
+    private RemoteIterator<LocatedFileStatus> remoteIterator;
+
+    public HiveFileIterator(
+            Path path,
+            FileSystem fileSystem,
+            DirectoryLister directoryLister,
+            NamenodeStats namenodeStats,
+            String partitionName,
+            InputFormat<?, ?> inputFormat,
+            Properties schema,
+            List<HivePartitionKey> partitionKeys,
+            TupleDomain<HiveColumnHandle> effectivePredicate)
+    {
+        this.partitionName = checkNotNull(partitionName, "partitionName is null");
+        this.inputFormat = checkNotNull(inputFormat, "inputFormat is null");
+        this.schema = checkNotNull(schema, "schema is null");
+        this.partitionKeys = checkNotNull(partitionKeys, "partitionKeys is null");
+        this.effectivePredicate = checkNotNull(effectivePredicate, "effectivePredicate is null");
+        this.path = checkNotNull(path, "path is null");
+        this.fileSystem = checkNotNull(fileSystem, "fileSystem is null");
+        this.directoryLister = checkNotNull(directoryLister, "directoryLister is null");
+        this.namenodeStats = checkNotNull(namenodeStats, "namenodeStats is null");
+    }
+
+    @Override
+    protected LocatedFileStatus computeNext()
+    {
+        try {
+            if (remoteIterator == null) {
+                remoteIterator = getLocatedFileStatusRemoteIterator(path);
+            }
+
+            while (remoteIterator.hasNext()) {
+                LocatedFileStatus status = getLocatedFileStatus(remoteIterator);
+
+                // ignore hidden files. Hive ignores files starting with _ and . as well.
+                String fileName = status.getPath().getName();
+                if (fileName.startsWith("_") || fileName.startsWith(".")) {
+                    continue;
+                }
+                return status;
+            }
+            return endOfData();
+        }
+        catch (FileNotFoundException e) {
+            throw new PrestoException(HiveErrorCode.HIVE_FILE_NOT_FOUND, "Partition location does not exist: " + path);
+        }
+        catch (IOException e) {
+            throw new PrestoException(HiveErrorCode.HIVE_UNKNOWN_ERROR, e);
```

Comment:
I think this is what the old code did, but we can do better since we know this is an `IOException`. Use either `HIVE_FILESYSTEM_ERROR` or perhaps a new `HIVE_DIRECTORY_LISTING_ERROR`?